opentracing-impl-java8/src/main/java/opentracing/propagation/AbstractTracer.java

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing.propagation;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import opentracing.Span;
+import opentracing.Tracer;
+
+abstract class AbstractTracer implements Tracer {
+
+    static final boolean BAGGAGE_ENABLED = !Boolean.getBoolean("opentracing.propagation.dropBaggage");
+
+    private final PropagationRegistry registry = new PropagationRegistry();
+
+    protected AbstractTracer() {
+        registry.register(Map.class, new SplitTextFormatInjectorImpl());
+        registry.register(Map.class, new SplitTextFormatExtractorImpl());
+    }
+
+    @Override
+    public Span startSpan(String operationName, /*@Nullable*/ Span parent) {
+        return startSpan(operationName, Optional.ofNullable(parent));
+    }
+
+    public Span startSpan(String operationName, Optional<Span> parent) {
+        return new SpanImpl(operationName, parent);
+    }
+
+    @Override
+    public <T> void inject(Span span, T carrier) {
+        registry.getInjector((Class<T>)carrier.getClass()).injectSpan(span, carrier);
+    }
+
+    @Override
+    public <T> Span join(String operationName, T carrier) {
+        return registry.getExtractor((Class<T>)carrier.getClass()).joinTrace(operationName, carrier);
+    }
+
+    public <T> Injector<T> register(Class<T> carrierType, Injector<T> injector) {
+        return registry.register(carrierType, injector);
+    }
+
+    public <T> Extractor<T> register(Class<T> carrierType, Extractor<T> extractor) {
+        return registry.register(carrierType, extractor);
+    }
+
+    private static class PropagationRegistry {
```

Comment:
in reality, such a tracing system would only support ^^ for interop reasons. In real life, the Injector/Extractor types would be injected directly into the call sites.

Ex. 
```java
// ignore the interfaces, and how the fields are initialized
class InjectCurrentSpanIntoMetadata implements Func1<Metadata> {
  Context.Key<Span> spanKey;
  Injector<Metadata> injector;

  void apply(Metadata input) {
    injector.injectSpan(spanKey.get(), input);
  }
}
```

Main thing is that the above type really needs no reference to a tracer to do its job, and it can be modified independently without impacting the tracer.